pkg/featflag/features.go

```diff
@@ -16,52 +16,198 @@
+// }
+// ```
+//
+// example 2: define a local feature flag
+// ```go
+// chefFeatXYZ := featflag.New("CHEF_FEAT_XYZ", "xyz")
+// if chefFeatXYZ.Enabled() {
+//   // the XYZ feature is enabled, act upon it
+// }
+// ```
+type Feature struct {
+	// the key associated to the feature flag defined inside the configuration file (config.toml)
+	//
+	// example of a config key:
+	// ```toml
+	// [features]
+	// analyze = true
+	// xyz = false
+	// ```
+	configKey string
+
+	// the environment variable name associated to the feature flag
+	//
+	// example of environment variables:
+	// CHEF_FEAT_ANALYZE=true
+	// CHEF_FEAT_XYZ=true
+	envName string
+}
+
+// global vs local feature flags
+//
+// when do I define a global feature flag?
+// > when that flag is being used by multiple packages
+//
+// NOTE: all environment variables and config keys are unique
+var (
+	// this special feature will enable all features at once
+	ChefFeatAll = Feature{
+		configKey: "all",
+		envName:   "CHEF_FEAT_ALL",
+	}
+
+	// enables the chef-analyze feature
+	ChefFeatAnalyze = Feature{
+		configKey: "analyze",
+		envName:   "CHEF_FEAT_ANALYZE",
+	}
+
+	// a list of all feature flags, global and local
+	featureFlags = []Feature{ChefFeatAll, ChefFeatAnalyze}
 
-const (
-	ChefFeatAll Feature = iota
-	ChefFeatAnalyze
+	// config instance to access feature keys
+	cfg *config.Config
 )
 
-var featureFlags = map[Feature]string{
-	ChefFeatAll:     "CHEF_FEAT_ALL",
-	ChefFeatAnalyze: "CHEF_FEAT_ANALYZE",
+func init() {
+	c, _ := config.New()
+	cfg = &c
 }
 
-func New(envName string) Feature {
-	feat := Feature(len(featureFlags) + 1 + int(ChefFeatAll))
-	featureFlags[feat] = envName
+// registers a new feature flag
+//
+// example of a new feature flag called 'foo':
+// ```go
+// chefFeatFoo := featflag.New("CHEF_FEAT_FOO", "foo")
+// chefFeatFoo.Enabled()  // returns true if the feature flag is enabled
+// ```
+func New(envName, key string) Feature {
+	// since all environment variables and config keys are unique,
+	// to protect them, this function will verify if there is a
+	// registered feature flag with any field, if so, it returns it
+	if feat, exist := GetFromEnv(envName); exist {
+		return *feat
+	}
+	if feat, exist := GetFromKey(key); exist {
+		return *feat
+	}
+
+	// create a new feature
+	feat := Feature{
+		configKey: key,
+		envName:   envName,
+	}
+
+	// register the new feature
+	featureFlags = append(featureFlags, feat)
+
 	return feat
 }
 
-func Get(envName string) (Feature, bool) {
-	for feat, name := range featureFlags {
-		if envName == name {
-			return feat, true
+// load a custom configuration instance,
+// this config is used inside the func 'Enabled()'
+func LoadConfig(c *config.Config) {
+	cfg = c
+}
+
+func ListAll() string {
+	list := make([]string, len(featureFlags))
+	for i, feat := range featureFlags {
+		list[i] = feat.String()
+	}
+
+	return strings.Join(list, " ")
+}
+
+func GetFromKey(key string) (*Feature, bool) {
```

Comment:
This is more a question on coding styles: what do you think about creating some maps to store the feature flags by key and environment name so the lookup time is O(1)? With a small number of expected flags it is a negligent execution time difference, but wanted to get your opinion on lookup time vs memory usage.

Reply:
I like the lookup func. 👍 Would you agree that, if we implement `lookup()` then the other two functions would go away? or have them also available? 🤔

Reply:
I think we would still need 2 functions since one would lookup by feature name and one would lookup by feature environment variable. Is `lookup` a standard function defined on slices? I'm fine with the current names but was more proposing a change to their implementation.

Reply:
I'll try to implement it, or we can pair on it. A few examples are here:

https://golang.org/src/os/env.go?s=3235:3276#L102
https://golang.org/src/net/lookup.go
https://golang.org/src/os/user/lookup.go